sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1405,7 +1405,8 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 case class RepartitionByExpression(
     partitionExpressions: Seq[Expression],
     child: LogicalPlan,
-    optNumPartitions: Option[Int]) extends RepartitionOperation {
+    optNumPartitions: Option[Int],
+    userSpecified: Boolean = true) extends RepartitionOperation {
```

Comment:
I'm afraid some rules will remove `RepartitionByExpression` without check the `userSpecified`, e.g. `RepartitionOperation` which may cause the output partitioning unexpected. That's why we add a new operator `RebalancePartitions` before. If the only difference is `optNumPartitions`, can't we just reuse `RebalancePartitions` ?

Reply:
good point. Maybe we should update `RebalancePartitions` to support `RangePartitioning`.

Reply:
A little confused here, doesn't the `Option` part of `optNumPartitions: Option[Int]` already contain the information of `userSpecified`?

Reply:
+1 on the idea of using `RebalancePartitions` instead. We should not mess up the semantics of `RepartitionByExpression`.

Reply:
I quickly tried `RebalancePartitions`, but it introduced one extra sort and shuffle. Is this OK?

Reply:
cc @aokolnychyi @RussellSpitzer